src/index.js

```diff
@@ -1,19 +1,252 @@
+  dialById (id, protocol, callback) {
+    // NOTE: dialById only works if a previous dial was made. This will
+    // change once we have PeerRouting
+
+    assert(this.isOnline, OFFLINE_ERROR_MESSAGE)
+
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    callback(new Error('not implemented yet'))
+  }
+
+  dialByMultiaddr (maddr, protocol, callback) {
+    assert(this.isOnline, OFFLINE_ERROR_MESSAGE)
+
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    if (typeof maddr === 'string') {
+      maddr = multiaddr(maddr)
+    }
+
+    if (!mafmt.IPFS.matches(maddr.toString())) {
+      return callback(new Error('multiaddr not valid'))
+    }
+
+    const ipfsIdB58String = maddr.stringTuples().filter((tuple) => {
+      if (tuple[0] === IPFS_CODE) {
+        return true
+      }
+    })[0][1]
+
+    let peer
+    try {
+      peer = this.peerBook.getByB58String(ipfsIdB58String)
+    } catch (err) {
+      peer = new PeerInfo(PeerId.createFromB58String(ipfsIdB58String))
+    }
+
+    peer.multiaddr.add(maddr)
+    this.dialByPeerInfo(peer, protocol, callback)
   }
 
-  if (!swarm) {
-    throw new Error('a swarm must be passed')
+  dialByPeerInfo (peer, protocol, callback) {
+    assert(this.isOnline, OFFLINE_ERROR_MESSAGE)
+
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    this.swarm.dial(peer, protocol, (err, conn) => {
+      if (err) {
+        return callback(err)
+      }
+      this.peerBook.put(peer)
+      callback(null, conn)
+    })
   }
 
-  self.swarm = swarm
-  self.routing = peerRouting
-  self.record = recordStore
+  //
+  // Disconnecting (hangUp) methods
+  //
+
+  hangUpById (id, callback) {
+    // TODO
+    callback(new Error('not implemented yet'))
+  }
+
+  hangUpByMultiaddr (maddr, callback) {
+    assert(this.isOnline, OFFLINE_ERROR_MESSAGE)
+
+    if (typeof maddr === 'string') {
+      maddr = multiaddr(maddr)
+    }
+
+    if (!mafmt.IPFS.matches(maddr.toString())) {
+      return callback(new Error('multiaddr not valid'))
+    }
+
+    const ipfsIdB58String = maddr.stringTuples().filter((tuple) => {
+      if (tuple[0] === IPFS_CODE) {
+        return true
+      }
+    })[0][1]
+
+    try {
+      const pi = this.peerBook.getByB58String(ipfsIdB58String)
+      this.hangUpByPeerInfo(pi, callback)
+    } catch (err) {
+      // already disconnected
+      callback()
+    }
+  }
+
+  hangUpByPeerInfo (peer, callback) {
+    assert(this.isOnline, OFFLINE_ERROR_MESSAGE)
+
+    this.peerBook.removeByB58String(peer.id.toB58String())
+    this.swarm.hangUp(peer, callback)
+  }
+
+  //
+  // Protocol multiplexing handling
+  //
+
+  handle (protocol, handlerFunc, matchFunc) {
+    this.swarm.handle(protocol, handlerFunc, matchFunc)
+  }
+
+  unhandle (protocol) {
+    this.swarm.unhandle(protocol)
+  }
+}
+
+module.exports = {
+  Node: Node
```

Comment:
Why not `module.exports = Node`? simpler to import and there is only one thing exported